In [1]:
## Updated to include labels from CCF v3
## CSV of labels from https://github.com/NeuroDataDesign/seelviz/blob/gh-pages/ccrf/ccrfCSV.csv

## UPDATED for spacing!

In [2]:
from plotly.offline import download_plotlyjs
from plotly.graph_objs import *
from plotly import tools
import plotly

import os
#os.chdir('C:/Users/L/Documents/Homework/BME/Neuro Data I/Data/')

import csv,gc  # garbage memory collection :)

import numpy as np
# import matplotlib.pyplot as plt
# from mpl_toolkits.mplot3d import axes3d

# from mpl_toolkits.mplot3d import axes3d
# from collections import namedtuple

import csv
import re
import matplotlib
import time
import seaborn as sns

from collections import OrderedDict

In [3]:
token = 'Aut1367'

In [4]:
"""Script for generating the color coded atlas region graphs"""

path1 = os.getcwd() + '/../ccrf/'
path2 = os.getcwd() + '/../ccf/'

data_txt = path1 + 'Aut1367reorient_atlas.region.csv'
ccf_txt = path2 + 'natureCCFOhedited.csv'

data = np.genfromtxt(data_txt, delimiter=',', dtype='int', usecols = (0,1,2,4), names=['x','y','z','region'])

In [5]:
# Debugging:
print data
# print type(ccf_txt)
# print type(data_txt)

[(37, 483, 292, 527) (37, 484, 304, 527) (39, 496, 330, 527) ...,
 (547, 466, 263, 527) (547, 470, 244, 527) (550, 470, 253, 527)]


In [5]:
ccf = {}

In [6]:
with open(ccf_txt, 'rU') as csvfile:
    csvreader = csv.reader(csvfile)
    for row in csvreader:
        # row[0] is ccf atlas index, row[4] is string of full name
        ccf[row[0]] = row[4];
        #print row[0]
        #print row[4]
        #print ', '.join(row)

In [8]:
# Debugging:
print ccf

{'10678': 'Crus 2, granular layer', '507': 'Main olfactory bulb', '198': 'pyramidal decussation', '1144': 'Cerebellar cortex, molecular layer', '1145': 'Cerebellar cortex, Purkinje layer', '161': 'Nucleus intercalatus', '344': 'Agranular insular area, posterior part, layer 5', '345': 'Primary somatosensory area, mouth', '346': 'Primary somatosensory area, layer 2/3', '347': 'Subparaventricular zone', '340': 'Posterior parietal association areas, layer 6b', '341': 'supramammillary decussation', '342': 'Substantia innominata', '343': 'Brain stem', '810': 'Anterior cingulate area, ventral part, 6a', '811': 'Inferior colliculus, central nucleus', '812': 'superior cerebellar peduncle decussation', '813': 'hypoglossal nerve', '814': 'Dorsal peduncular area', '815': 'Dentate gyrus medial blade, polymorph layer', '816': 'Primary auditory area, layer 4', '817': 'supraoptic commissures, anterior', '380': 'cuneate fascicle', '595': 'fasciculus retroflexus', '737': 'postcommissural fornix', '719':

In [9]:
# Sanity check:
print ccf['214']
print ccf['1']

Red nucleus
Tuberomammillary nucleus, ventral part


In [10]:
# Sanity check:
print type(data)
print data[3]

<type 'numpy.ndarray'>
(41, 505, 370, 527)


In [7]:
# Find all unique regions of brightest points
unique = [];

for l in data:
    unique.append(l[3])

uniqueNP = np.asarray(unique)
allUnique = np.unique(uniqueNP)
numRegionsA = len(allUnique)

print allUnique
print numRegionsA  ## number of regions

[   1    2    7    8    9   10   12   15   17   19   20   23   26  257  258
  259  260  262  263  266  267  268  269  271  272  274  276  278  279  280
  281  513  515  520  521  523  524  526  527  529  531  532  534  537  769
  771  772  773  774  777  778  780  781  783  785  786  791 1026 1029 1030
 1031 1034 1035 1038 1042 1044 1045 1046 1047 1048 1050]
71


In [8]:
# Store and count the bright regions in each unique region
dictNumElementsRegion = {}

for i in range(numRegionsA):
    counter = 0;
    for l in data:
        if l[3] == allUnique[i]:
            counter = counter + 1;
            dictNumElementsRegion[ccf[str(l[3])]] = counter;

In [9]:
region_names = dictNumElementsRegion.keys()
number_repetitions = dictNumElementsRegion.values()

In [14]:
print dictNumElementsRegion

{'Superior colliculus, motor related, intermediate gray layer': 352, 'Bed nuclei of the stria terminalis, anterior division, fusiform nucleus': 2, 'Primary somatosensory area, upper limb, layer 6b': 31, 'Posterior parietal association areas, layer 1': 183, 'posteromedial visual area, layer 6a': 46, 'Superior colliculus, motor related, deep gray layer': 324, 'Taenia tecta, dorsal part, layer 2': 50, 'Supratrigeminal nucleus': 48, 'Posterolateral visual area, layer 2/3': 48, 'Taenia tecta, dorsal part, layer 1': 37, 'Anterior amygdalar area': 177, 'Principal sensory nucleus of the trigeminal': 327, 'Posterior amygdalar nucleus': 93, 'Orbital area, medial part, layer 2': 7, 'Lateral septal nucleus, ventral part': 14, 'Dorsal auditory area, layer 1': 65, 'Taenia tecta, dorsal part, layer 3': 33, 'Primary somatosensory area, barrel field, layer 6a': 395, 'Parataenial nucleus': 24, 'Retrosplenial area, lateral agranular part, layer 6b': 7, 'Reticular nucleus of the thalamus': 271, 'Primary s

In [15]:
print region_names
print number_repetitions

['Superior colliculus, motor related, intermediate gray layer', 'Bed nuclei of the stria terminalis, anterior division, fusiform nucleus', 'Primary somatosensory area, upper limb, layer 6b', 'Posterior parietal association areas, layer 1', 'posteromedial visual area, layer 6a', 'Superior colliculus, motor related, deep gray layer', 'Taenia tecta, dorsal part, layer 2', 'Supratrigeminal nucleus', 'Posterolateral visual area, layer 2/3', 'Taenia tecta, dorsal part, layer 1', 'Anterior amygdalar area', 'Principal sensory nucleus of the trigeminal', 'Posterior amygdalar nucleus', 'Orbital area, medial part, layer 2', 'Lateral septal nucleus, ventral part', 'Dorsal auditory area, layer 1', 'Taenia tecta, dorsal part, layer 3', 'Primary somatosensory area, barrel field, layer 6a', 'Parataenial nucleus', 'Retrosplenial area, lateral agranular part, layer 6b', 'Reticular nucleus of the thalamus', 'Primary somatosensory area, lower limb, layer 1', 'Entorhinal area, lateral part, layer 2', 'Vent

In [16]:
from itertools import izip

with open('Aut1367Counts.csv', 'wb') as write:
    writer = csv.writer(write)
    writer.writerows(izip(region_names, number_repetitions))

Since we know the dimensions from GetSpacing on the img (run on the Google server) we know: 

Aut1367:  
Raw is X, Y, Z:  
shape = (595, 912, 1225)  
resolution = (0.01872, 0.01872, 0.005)  
(and do we need this?)  
Final is Z, Y, X  
(1225, 912, 595)  
(1.0, 1.0, 1.0)  

Thus I scale each x, y by multiplying w/ 0.01872 to get mm accuracy.  
I do the same for z.

In [10]:
def generate_atlas_region_graph(data, path=None, numRegions = 71):
    font = {'weight' : 'bold',
        'size' : 18}

    matplotlib.rc('font', **font)

    region_dict = OrderedDict()

    for l in data:

        trace = ccf[str(l[3])]
        #trace = 'trace' + str(l[3])
        if trace not in region_dict:
            region_dict[trace] = np.array([[l[0], l[1], l[2], l[3]]])
            #print 'yay'
        else:
            tmp = np.array([[l[0], l[1], l[2], l[3]]])
            region_dict[trace] = np.concatenate((region_dict.get(trace, np.zeros((1,4))), tmp), axis=0)
            #print 'nay'

    current_palette = sns.color_palette("husl", numRegions)
    # print current_palette

    data = []
    for i, key in enumerate(region_dict):
        trace = region_dict[key]
        tmp_col = current_palette[i]
        tmp_col_lit = 'rgb' + str(tmp_col)
        temp = str(np.unique(trace[:,3])).replace("[", "")
        final = temp.replace("]", "")
    
        trace_scatter = Scatter3d(
            x = trace[:,0] * 0.01872, 
            y = trace[:,1] * 0.01872,
            z = trace[:,2] * 0.005,
            mode='markers',
            name=ccf[final],
            marker=dict(
                size=1.2,
                color=tmp_col_lit, #'purple',                # set color to an array/list of desired values
                colorscale='Viridis',   # choose a colorscale
                opacity=0.15
            )
        )

        data.append(trace_scatter)


    layout = Layout(
        margin=dict(
            l=0,
            r=0,
            b=0,
            t=0
        ),
        paper_bgcolor='rgb(0,0,0)',
        plot_bgcolor='rgb(0,0,0)'
    )

    fig = Figure(data=data, layout=layout)
    plotly.offline.plot(fig, filename= 'Aut1367' + "_region_color_spaced.html")

In [18]:
generate_atlas_region_graph(data, path=None, numRegions = numRegionsA);

In [13]:
    region_dict = OrderedDict()

    for l in data:

        trace = ccf[str(l[3])]
        #trace = 'trace' + str(l[3])
        if trace not in region_dict:
            region_dict[trace] = np.array([[l[0], l[1], l[2], l[3]]])
            #print 'yay'
        else:
            tmp = np.array([[l[0], l[1], l[2], l[3]]])
            region_dict[trace] = np.concatenate((region_dict.get(trace, np.zeros((1,4))), tmp), axis=0)
            #print 'nay'

    current_palette = sns.color_palette("husl", numRegionsA)
    # print current_palette

    data = []
    for i, key in enumerate(region_dict):
        trace = region_dict[key]
        tmp_col = current_palette[i]
        tmp_col_lit = 'rgb' + str(tmp_col)
        temp = str(np.unique(trace[:,3])).replace("[", "")
        final = temp.replace("]", "")
    
        trace_scatter = Scatter3d(
            x = trace[:,0] * 0.01872, 
            y = trace[:,1] * 0.01872,
            z = trace[:,2] * 0.005,
            mode='markers',
            name=ccf[final],
            marker=dict(
                size=1.2,
                color=tmp_col_lit, #'purple',                # set color to an array/list of desired values
                colorscale='Viridis',   # choose a colorscale
                opacity=0.15
            )
        )

        data.append(trace_scatter)

In [20]:
print trace[:,0]
print trace[:,0] * 0.01872

[343 347]
[ 6.42096  6.49584]
